### PR TITLE
Make certain types more polymorphic

### PR DESCRIPTION
**Pull request by adinapoli-mndc**
In my quest to fix osrank-rs#85, I have generalised some types to accept a polymorphic `W`, so that we can instantiate this on the `osrank-rs` side with either a proper fractional type (for testing) for with a `f64` (for production).

On top of that, I have also added a bunch of `Arbitrary` instances which we need in the tests.

The most controversial change might be the changes to the `EdgeType` type, which has now been modelled not in terms of the "entity-relational" high-level schema, but in terms of the "low-level" osrank representation. There are two reasons behind this change:

1. The previous `edge_weights: HashMap<EdgeType, f64>` was incorrect, because the paper defines 5 distinct weights (`depend`, `contrib`, `contrib*`, `maintain`, `maintain*`) but due to the fact the cardinality of the previous `EdgeType` was 3, we couldn't store all the hyperparams correctly;

2. We can always recover the previous notion of "high-level types" by having simple helper functions which can be called from the ledger to set the correct edge type which can get "desugared" into the low-level `EdgeType`. Practical example: we could have `add_edge_foobar(n1: Node, n2: Node, ...)` which would look at the `NodeType` of `n1` and `n2` and determine the right edge type.

Early feedback welcome! I should also say that all the changes I am making here and being driven by `osrank-rs`.

### CODE REVIEW

**Comment by adinapoli-mndc** [comment]
This is a starting point. Some choices might be controversial and not particularly orthodox, but this is a starting point which is proved to work in the context of `osrank`.

File: examples/main.rs [hunk 1]
```diff
@@ -19,9 +20,14 @@ type EdgeData = Bytes;
 pub struct Node {
     id: Id,
     data: NodeData,
+    node_type: types::NodeType,
```
Comment (cloudhead) [inline examples/main.rs]: Why is this separate from `NodeData`?

File: examples/main.rs [hunk 2]
```diff
@@ -30,12 +36,25 @@ pub struct Edge {
     to: Id,
     data: EdgeData,
     weight: f64,
+    edge_type: types::EdgeType,
```
Comment (cloudhead) [inline examples/main.rs]: Same question here

File: src/lib.rs [hunk 3]
```diff
@@ -86,6 +97,15 @@ pub trait GraphWriter: Graph + GraphDataWriter {
     fn nodes_mut(&mut self) -> NodesMut<Self::Node>;
 }
 
+/// A graph with read-only access to edge and node data.
+pub trait GraphDataReader: Graph {
+    /// Return an immutable reference to an edge's data.
+    fn edge_data(&self, id: &Id<Self::Edge>) -> Option<&Data<Self::Edge>>;
+
+    /// Return an immutable reference to a node's data, to annotate the node.
```
Comment (cloudhead) [inline src/lib.rs]:
```suggestion
    /// Return an immutable reference to a node's data.
```

File: src/types.rs [hunk 4]
```diff
@@ -1,51 +1,154 @@
...
+}
+
 /// Node data.
-type NodeData = NodeType;
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct NodeData<W> {
+    /// The type for this node.
+    pub node_type: NodeType,
+    /// The total contributions by this user, to *all* projects, if any.
+    pub total_contributions: Option<u32>,
```
Comment (cloudhead) [inline src/types.rs]: Hm?

File: src/types.rs [hunk 5]
```diff
@@ -1,51 +1,154 @@
...
 pub enum EdgeType {
-    /// Contribution from a user to a project. This edge type is
-    /// used bi-directionally, between project and user.
-    Contribution,
-    /// Membership of a user in a project, project <-> user.
-    ProjectMembership,
+    /// Contribution from a project to a user.
+    Contrib,
+    /// Contribution from a user to a project.
+    ContribStar,
```
Comment (cloudhead) [inline src/types.rs]: Can we call this something meaningful?

File: src/types.rs [hunk 6]
```diff
@@ -1,51 +1,154 @@
...
+    /// The type for this edge.
+    pub edge_type: EdgeType,
     /// The weight of this specific edge. Can be used to weight for eg.
     /// edges with more contributions higher, or weigh certain dependencies
-    /// higher than others. Defaults to `1.0`.
-    weight: f64,
+    /// higher than others.
+    pub weight: W,
+    /// The contributions of the user towards a project, if any.
+    pub contributions: Option<u32>,
```
Comment (cloudhead) [inline src/types.rs]: Hm?

**Comment by cloudhead** [comment]
Regarding 1., see: https://github.com/oscoin/graph-api/pull/6#pullrequestreview-287939959

File: Cargo.toml [hunk 7]
```diff
@@ -3,3 +3,8 @@ name = "oscoin-graph-api"
 version = "0.1.0"
 authors = ["Alexis Sellier <alexis@monadic.xyz>"]
 edition = "2018"
+
+[dependencies]
+
+quickcheck = "*"
```
Comment (cloudhead) [inline Cargo.toml]: Can this be a `dev-dependencies` perhaps?

Comment (adinapoli-mndc) [reply Cargo.toml]: I am not a Cargo-toml-master, but my understanding is that you can do that if you are relying on `quickcheck` only for test code (or at least this maps from my intuition coming from the Haskell world). Here, however, we are defining a bunch of `Arbitrary` instances for library code, which I would have thought it warrants the import of `quickcheck` as a normal dependency. Or are you saying that in principle one can do something like this, in library code?

```rust

#[cfg(test)]
impl Arbitrary for Foo {
...
}
```

File: src/types.rs (hunk 5 above)
Comment (adinapoli-mndc) [reply src/types.rs]: I have tried to use a naming scheme as close as possible to the way the paper also describes the edge types. In the paper we have this unicode "circle" symbol, like in `contrib⚬`, which I didn't know how to vocalise in plain english. Usually when you want to express the dual for something you tend to also use "star" (as in `contrib*`) or "prime" (as in `contrib'`). I am personally neutral -- I have opted for the former, but I am open to suggestions.

File: examples/main.rs (hunk 1 above)
Comment (adinapoli-mndc) [reply examples/main.rs]: Ah, simply for conveniency -- this example is using blob of bytes as our `NodeData`, but really what I wanted to use here was the concrete node data, but I also didn't want to throw away the part of the code you did write. Suggestions? I am totally fine with having `NodeData` be `types::NodeData`, and then remove this `node_type` field.

File: examples/main.rs (hunk 2 above)
Comment (adinapoli-mndc) [reply examples/main.rs]: Some reply as above 😛

File: src/types.rs (hunk 4 above)
Comment (adinapoli-mndc) [reply src/types.rs]: This is an attempt to model (dynamic) contributions data for a project or an account, so that we can rely on these when we compute the osrank, see [here](https://github.com/oscoin/osrank-rs/pull/87/files#diff-88175c33d601b2ae4ee5eb28e8f2b549R222-R237). 

As regards why this has been modelled as an `Option` (if your confusion stemmed from that), I am not fully satisfied either. What I wanted to express here is that there are situations where you have no contributions, but at the same time you don't want to use `0`, as that might trigger divisions by 0 down the line and you would have to carefully check for that. However, this now moves the burden of constructing semantically-correct graphs into the registry/graph api layer, as for example having `Some(0)` would be wrong.

I am open to suggestions! Should we simply go with plain `u32`, here?

File: src/types.rs (hunk 6 above)
Comment (adinapoli-mndc) [reply src/types.rs]: See above! 😉

**Comment by adinapoli-mndc** [comment]
> Regarding 1., see: #6 (review)

@cloudhead I am not sure I follow/necessarily agree with the conclusions in that discussion, heh. It's very true that we can do work to infer the type of the edge by looking at `(source,target)` 's `NodeType`'s, and it is also true this will save us some precious bytes in memory as it will now mean that every time we add a `Contrib/Maintain` edge this automatically means that you have a dual going into the opposite direction (@MeBrei correct me if I am wrong, but I don't think you can ever find yourself into a situation where you have a `Contrib/Maintain` without the dual `Contrib*/Maintain*`).

Having said that, this wouldn't solve the `HashMap` lookup problem, right? If I have now only 3 keys to store the edge weights, namely `Contrib, Depend, Maintain`, how can I get the hyperparam weights for `Contrib*` and `Maintain*`? Note that they all have different values, at the moment of writing:

```
contrib: 1/7
contrib*: 2/5
maintain: 2/7
maintain*: 3/5
depend: 4/7
``` 

This is why I have eventually settled on that `EdgeType` enum, and the osrank implementation followed from that. We just need to be aware that if we now revert to that simpler `EdgeType` enum, we would have also to amend the osrank code to deal with that, which might or might not be easy (I haven't tried yet, but coming up with the current [PR](https://github.com/oscoin/osrank-rs/pull/87) took me already some time).

File: Cargo.toml (hunk 7 above)
Comment (cloudhead) [reply Cargo.toml]: That's what I mean :relaxed:

**Comment by cloudhead** [comment]
> Having said that, this wouldn't solve the HashMap lookup problem, right?

Indeed, that's a good reason to have the individual edge types.

File: src/types.rs (hunk 5 above)
Comment (cloudhead) [reply src/types.rs]: It's `co-contrib` :relaxed: 

But I think I'd prefer something readable for someone who hasn't read the paper in detail, so maybe the names I had in https://github.com/oscoin/graph-api/pull/6#pullrequestreview-287939959 ?

File: Cargo.toml (hunk 7 above)
Comment (adinapoli-mndc) [reply Cargo.toml]: Ah, cool, didn't know Rust supported that, will give it a shot!

File: src/types.rs (hunk 5 above)
Comment (cloudhead) [reply src/types.rs]:
```
    /// Contribution, user -> project, with strength metric.
    ContributionToProject(u32),
    /// Contribution, project -> user, with strength metric.
    ContributionFromUser(u32),
```

Comment (adinapoli-mndc) [reply src/types.rs]: Yup, not a problem -- as long as there exist a 1:1 correspondence with the edge types in the paper, I am happy to use "developer-friendly" types ;)

File: examples/main.rs (hunk 1 above)
Comment (cloudhead) [reply examples/main.rs]: Yeah I'd just say add `type NodeData = NodeType` for eg.

File: examples/main.rs (hunk 2 above)
Comment (cloudhead) [reply examples/main.rs]: Same reply to your reply as above :stuck_out_tongue_closed_eyes:

File: src/types.rs (hunk 5 above)
Comment (adinapoli-mndc) [reply src/types.rs]:
```
  /// Contribution, user -> project, with strength metric.
    ContributionToProject(u32),
    /// Contribution, project -> user, with strength metric.
    ContributionFromUser(u32),
```

That's not gonna work my friend, as if you embed data into this enum, you cannot use this as a key in the `edge_weights` `HashMap` 😉

File: src/types.rs (hunk 4 above)
Comment (cloudhead) [reply src/types.rs]: Ah I see - my confusion is partly that not all nodes are users, but the comment says "total contribs by this user". I'd expect this to be part of node_data, with node_type, eg:

```rust
type NodeData = NodeType;
enum NodeType {
  Contributor({ contribs: u32 }),
  Project,
}
```

Does this make sense?

Comment (adinapoli-mndc) [reply src/types.rs]: > Does this make sense?

At 5PM of a Friday I will say "yes", but I will have to try this in the osrank code to check whether or not there are any blockers. Will give this a go on Monday and report back. "Stay tuned" 😁

Comment (adinapoli-mndc) [reply src/types.rs]: @cloudhead Ok, after a good restful weekend I can debunk my own statement 😂 : Unfortunately we do need to have a `contribs: Option<u32>` for *both* node types, because in the case of a user that will be (cfr. whitepaper) "the sum of all the contributions this user have done", whereas in the case of a project, this number will be "the sum of all the contributions it received from _all_ accounts".

My suggestion is that we adjust the comment, making it less confusing, but we stick with my original design. What we can reason about though is the use of `Option`, as I have said. Originally I did want to use `Option` to model the fact when a user joins the network it might not have contributed to anything, and ditto for a project, so using `None` would model the "lack of relevant data here". Perhaps using a simple `u32` would work better here (so that we can use `0` as the unit value here), provided we do our due diligence in `osrank` to make sure we always check for division by zero.

Thoughts?

**Comment by adinapoli-mndc** [comment]
@cloudhead Ok, modulo the discussion around where and how to represent contributions, I have applied all the other suggestions. Notable points:

* Naming is hard, so while I have tried to reuse the original naming scheme for `EdgeType`, there might be some names we might want to improve. I've tried to keep the type (i.e. contrib/maintain/...) always appear as a prefix, to make programming overhead lower. Due to the fact we are now deviating from the paper names, it's already harder in `osrank` to map the paper names to the new ones, when writing code 😉 

* The examples have been simplified by having `type NodeData = types::NodeData` and `type EdgeData = types::EdgeData` as you requested. This meant I had to simplify some parts of the examples, notably the checkpointing part, as now the `version` and `hash` do not fit into the current `NodeData` `struct`. I think in general this suggest we should add them at some point, but let's not go there as part of this PR and rather let's do this "on demand" as soon as the registry needs this stuff 😁 

Let me know what you think!

**Comment by adinapoli-mndc** [comment]
Damn, just realised I have made all my commits with my [doppelganger](https://en.wikipedia.org/wiki/Doppelg%C3%A4nger). Let me rewrite history 😅

File: src/types.rs (hunk 4 above)
Comment (cloudhead) [reply src/types.rs]: :thinking: 
Two things:
1. If we want contributions for both, I would still consider the sum-type, since the meaning of the field is not the same. I think it would be more confusing to have a single thing mean two things in different contexts.
2. I'm not sure why we need the "sum of contributions this user has done"?

File: src/types.rs (hunk 5 above)
Comment (cloudhead) [reply src/types.rs]: Yeah that's fine - either way you need two sets of data: one for the tag and one for the content (strength)

Comment (cloudhead) [reply src/types.rs]: What I mean by "either way" is you could keep that enum, and have another one with the tag only:

```rust
enum Foo { ContribTo(u32), ContribFrom(u32) }
impl Foo { fn tag(&self) -> FooTag { ... } }
enum FooTag { ContribTo, ContribFrom }
```

File: Cargo.toml (hunk 7 above)
Comment (cloudhead) [reply Cargo.toml]: I've never tried it tbh, but it *should* work :stuck_out_tongue_winking_eye:

File: src/lib.rs [hunk 8]
```diff
@@ -36,12 +39,21 @@ pub trait GraphObject {
 }
 
 /// A graph  node.
-pub trait Node<N>: GraphObject<Data = N> {}
+pub trait Node<N>: GraphObject<Data = N> {
+    /// Returns the type of this node.
+    fn node_type(&self) -> &types::NodeType;
```
Comment (cloudhead) [inline src/lib.rs:52]: This shouldn't be needed - the node already has data

File: src/lib.rs [hunk 9]
```diff
@@ -36,12 +39,21 @@ pub trait GraphObject {
...
-pub trait Edge<W, E>: GraphObject<Data = E> {
+pub trait Edge<W, NodeId, E>: GraphObject<Data = E> {
+    /// The source node.
+    fn source(&self) -> &NodeId;
+    /// The target node
+    fn target(&self) -> &NodeId;
     /// Get the edge weight.
     fn weight(&self) -> W;
+    /// Returns the type of this edge.
+    fn edge_type(&self) -> &types::EdgeType;
```
Comment (cloudhead) [inline src/lib.rs:65]: Same here

File: src/lib.rs [hunk 10]
```diff
@@ -230,6 +250,7 @@ pub struct EdgeRef<'a, NodeId, EdgeId> {
     pub from: &'a NodeId,
     pub to: &'a NodeId,
     pub id: &'a EdgeId,
+    pub edge_type: &'a EdgeType,
```
Comment (cloudhead) [inline src/lib.rs:271]: Should be `edge_data: EdgeData`

File: src/types.rs [hunk 11]
```diff
@@ -1,51 +1,160 @@
...
 /// The type of a node.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum NodeType {
     /// A user, eg. contributor, project member etc.
     User,
     /// A project with users as members and contributors.
     Project,
 }
 
+#[cfg(feature = "quickcheck")]
```
Comment (cloudhead) [inline src/types.rs:69]: Nice!

File: src/types.rs (hunk 4 above)
Comment (adinapoli-mndc) [reply src/types.rs]: > If we want contributions for both, I would still consider the sum-type, since the meaning of the field is not the same. I think it would be more confusing to have a single thing mean two things in different contexts.

Each design has pro & cons. I agree that it might be confusing to have the same field mean two different things, but it's also true that the context is set by the `NodeType`, i.e. when you pattern match on the node type, you will know which is the semantic of this field. Conversely, if we now convert the enum into a sum type, it means we cannot set the `NodeType` of a node without having to also know, apriori, the contributions. Perhaps this isn't a problem, but surely something to think about.

Something I like of your sum-type proposal instead is that paradoxically now it's harder to change the node type (due to the fact this is a sum type), which is kinda nice because it forces you to really bundle these two things together. I will play with your proposal and see if there is any showstopper on the osrank side.

> I'm not sure why we need the "sum of contributions this user has done"?

This is necessary for the weights calculation and for choosing the next edge of the random walk -- See for example section 3.5 of the paper, where we compute the probability of `contrib*` and `maintain*` by doing `px/Nx` where `Nx` is "the total number of contributions by x`, where `x` is a user.

File: src/types.rs [hunk 12]
```diff
@@ -1,51 +1,160 @@
...
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum EdgeType {
-    /// Contribution from a user to a project. This edge type is
-    /// used bi-directionally, between project and user.
-    Contribution,
-    /// Membership of a user in a project, project <-> user.
-    ProjectMembership,
-    /// One-way dependency between two projects, project -> project.
-    ProjectDependency,
+    /// Contribution from a project to a user. Corresponds to `contrib` from the paper.
```
Comment (cloudhead) [inline src/types.rs:110]:
```suggestion
    /// Contribution from a user to a project. Corresponds to `contrib` from the paper.
```

File: Cargo.toml (hunk 7 above)
Comment (adinapoli-mndc) [reply Cargo.toml]: Oh, it does, I have tried already. Interestingly enough what didn't work was your suggested approach (😭) and apparently I wasn't alone in my struggles:

* https://users.rust-lang.org/t/implementing-quickcheck-arbitrary-only-when-testing-crates/12851

* https://stackoverflow.com/questions/36390808/how-to-create-a-quickcheck-arbitrary-of-a-struct-containing-a-reference

Luckily the solution suggested in the first link (which incidentally is the one I have used) worked like a charm 😉

**Comment by cloudhead** [comment]
Looking good - just a few nits, and need to clarify this "all contribs from a user u32"

File: src/lib.rs (hunk 8 above)
Comment (adinapoli-mndc) [reply src/lib.rs:52]: This is needed because we are implementing the `Node` _trait_ here, and remember that the whole point of this exercise was to allow test code to swap the concrete `NodeData` for something more flexible/simpler. Therefore, it's user code's responsibility to implement this method correctly 😄 (see for example [my open osrank PR](https://github.com/oscoin/osrank-rs/pull/87/files#diff-a0e64c39037b366c6e86d7aa93feb56cR273), where we have an extra layer of indirection, for example).

File: src/lib.rs (hunk 10 above)
Comment (adinapoli-mndc) [reply src/lib.rs:271]: Why? If we use `EdgeData` here we are locking away the possibility to override the `EdgeData` with something simpler/richer which also _contains_ the `EdgeType`. If you remember from our previous discussions we settled on having `EdgeType` and `NodeType` being the lowest common denominator between the registry & osrank, but we also discussed how it's nice for osrank to be able to have full control on how to map this data.

My vote would be to leave `EdgeType` here, as it's literally all we need 😉

Comment (adinapoli-mndc) [reply src/lib.rs:271]: ps. Note how we cannot use `G::EdgeData` as we don't have a `G` in scope, alas, and due to the fact `EdgeType` was all I needed .. this is what I have added 😅

File: src/lib.rs (hunk 8 above)
Comment (cloudhead) [reply src/lib.rs:52]: Oh, then shouldn't it be eg. `node_data(&self) -> Self::Data`? Since `types` is the concrete types.

File: src/types.rs (hunk 12 above)
Comment (adinapoli-mndc) [reply src/types.rs:110]: Argh. I still find these new names still quite confusing (pardon me!). Let me have a sanity check here:

- `ContributionsFromUser` means that I am a project and I have an incoming edge from a
   user. So yeah, not only is your suggestion correct, but this is *not* a `contrib` edge, this
   is a `contrib*` edge.

- `ContributionToProject` is what trips me up, because that "to" suggest that I have an arrow
   going _into_ a project node, but this is not the case, this is what used to be a `contrib` edge
   and therefore has direction `project -> user`.

I think this is a recipe for disaster because if I did get them wrong with the paper in front of me, good luck trying to get these right in the wild 😅 

May I be pragmatic and propose the following mapping?

```rust
pub enum EdgeType {
    /// Contribution from a project to a user. Corresponds to `contrib` from the paper.
    ProjectToUserContribution,
    /// Contribution from a user to a project. Corresponds to `contribᵒ` from the paper.
    UserToProjectContribution,
    /// Membership relation from a project to a user. Corresponds to `maintain` from the paper.
    ProjectToUserMembership,
    /// Membership relation from a user to a project. Correspond to `maintainᵒ` from the paper.
    UserToProjectMembership,
    /// One-way dependency between two projects. Correspond to `depend` from the paper.
    Dependency,
}
```

Maybe this isn't glamorous, but it has the advantage that it states in the name of each enum the direction, and it's now very clear which is the direction of the arrow. What do you think?

File: src/types.rs (hunk 4 above)
Comment (cloudhead) [reply src/types.rs]: Yes, the whole point of the sum-type is to not mistakenly mix them up, and ensure that the data is only in scope when the node type is correct (statically).

As for the contribs thing, I'm fairly sure we changed this - I remember talking to @MeBrei and maybe also you about simplifying this logic, since it doesn't make that much sense. I'll try to find any trace of this discussion :thinking:

File: Cargo.toml (hunk 7 above)
Comment (cloudhead) [reply Cargo.toml]: Oh strange - I don't quite understand why `#[cfg(test)]` doesn't work, but your solution is :+1:

File: src/lib.rs (hunk 8 above)
Comment (adinapoli-mndc) [reply src/lib.rs:52]: There is a bit of history here -- originally I did code everything under the impression that the only little piece of concrete data we were allowing in the interfaces were the `types::NodeData` and `types::EdgeData`, so that both the registry and osrank could have relied on this during data-agnostic code. For example, this would have allowed us to write a generic algorithm function where you had only a `G: Graph` constraint while being able to pattern match on each node type by the virtue of relying on the (concrete) info that each (abstract) node had such a `node_type` function.

However, for other reasons I had to make our osrank algorithm function partially monomorphic in the `G::Data`, which means that, short story short, I should be able to kill this `node_type` method completely, because we can already grab the data data using the `GraphDataReader` or, at the node-level, simply relying on the `GraphObject.data()` function.

Let me have a play with this.

File: src/types.rs (hunk 4 above)
Comment (adinapoli-mndc) [reply src/types.rs]: > As for the contribs thing, I'm fairly sure we changed this - I remember talking to @MeBrei and maybe also you about simplifying this logic, since it doesn't make that much sense. I'll try to find any trace of this discussion 🤔

As far as I know (but Merle can debunk me here) what changed is only the `maintain*` weight calculation, which is now equiprobable, whereas for `contrib*` we still need to do the `px/Nx` ratio, and thus you do need the "total number of contributions by x`.

File: src/types.rs (hunk 12 above)
Comment (adinapoli-mndc) [reply src/types.rs:110]: Let me be proactive and change this to the proposed enumeration -- we can always revert in case 😉 (EDIT: https://github.com/oscoin/graph-api/pull/8/commits/f9f475f8bd7d59b6e07b536a43d2687ee4212af6)

Comment (cloudhead) [reply src/types.rs:110]: Yeah I think this is the way to go :+1:

File: src/types.rs (hunk 4 above)
Comment (MeBrei) [reply src/types.rs]: > As far as I know (but Merle can debunk me here) what changed is only the maintain* weight calculation, which is now equiprobable, whereas for contrib* we still need to do the px/Nx ratio, and thus you do need the "total number of contributions by x`.

Yes, for `contrib` and `contrib*` we need the total amount of contributions for the node. There might be ways to work around this, although they have some downfalls:
1) We calculate the total on the fly ie we write a function that goes over each edge of the same type and sums the contributions or after we created the list of contributions, we simply sum it up.
2) We do not use the `total number of ...`. Currently, we use `choose_weighted`, to choose from the `contrib(*)` edges. The function automatically normalizes the weights it chooses from, so it will indirectly devide by the sum of all contributions. The indirectness might make bugs less obvious though.

@adinapoli-mndc am I missing another use of the `total number of ...`?
